### PR TITLE
New Attachment Code

### PR DESCRIPTION
**Pull request by SouthyOsmanli**
## About The Pull Request

Since I was unable to port most of Eris' tool/weapon code over to our codebase due to define differences and the like - I opted instead to look to other codebases for attachment code. 

Shiptest had a interesting attachment code PR by ZephrTFA ( https://github.com/shiptest-ss13/Shiptest/pull/145 ) done for them. The code looked interesting and worked on our codebase due to it being designed for a similar TG codebase. All credit for all the define and base attachments to them as they did the work on all that; I simply added our attachments and some balance to our codebase.

**PR is basically good to go.** There's some issues with removing attachments since I can't figure out how to fix it but, well, currently you can't remove them anyway I suppose so no real loss.

## Why It's Good For The Game

This **completely** redoes our attachment system. Attachments will no longer be straight upgrades; but rather trade-offs.

Examples are:
- Autosear and Recoil Comps now consume the same attachment slot. So if you can make a gun automatic it will be unable to take a recoil compensator while it is automatic.
- The extended barrel adds 1 fire delay in exchange for bullet projectile speed.
- Laser sights and rail lights take up the same slot, as do a suppressor vs a bayonet.
This will work well with the gun-balance changes to stop people from making tons of OP as fuck weapons with no downsides by spending 100 caps on each. This allows guns to serve specific purposes or allow for different build styles on firearms with tradeoffs to each.

## Changelog
:cl:
add: New attachment system.
add: New attachments for both ballistics and AER-9, AER-12 and AER-14 platforms.
balance: Attachment values have been changed across the board; though roughly do the same thing.
/:cl:

### CODE REVIEW

File: code/datums/components/attachment.dm [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+#define COMSIG_ATTACHMENT_DETACH "attach-detach"
+#define COMSIG_ATTACHMENT_EXAMINE "attach-examine"
+#define COMSIG_ATTACHMENT_EXAMINE_MORE "attach-examine-more"
+#define COMSIG_ATTACHMENT_PRE_ATTACK "attach-pre-attack"
+#define COMSIG_ATTACHMENT_ATTACK "attach-attacked"
+#define COMSIG_ATTACHMENT_UPDATE_OVERLAY "attach-overlay"
+
+//Basically the parts of guns where shit attaches to.
+#define COMSIG_ATTACHMENT_GET_SLOT "attach-slot-who"
+	#define ATTACHMENT_SLOT_MUZZLE "attach-slot-muzzle"
```
Comment (Trilbyspaceclone) [inline code/datums/components/attachment.dm:11]: Why are all these tabbed in once?

Comment (SouthyOsmanli) [reply code/datums/components/attachment.dm:11]: Probably error on my part when bringing the code over. Didn't seem to cause an issue (?)

**Comment by SouthyOsmanli** [comment]
done developing for the server at least for now.

**Comment by SouthyOsmanli** [comment]
Working on this again but still finding issues. Though I've figured out half the issues.

**Pull request closed without merging**